### PR TITLE
Lab 2: Bubble Sort

### LINKED ISSUE #540: Lab 2 - Bubble sort
Write a function that returns a sorted copy of int slice s using bubble sort:
func bubble(s []int) []int
Call fmt.Println(bubble([]int{3, 2, 1, 5})) in main to print:
[1 2 3 5]

### PR DESCRIPTION
**Pull request by krnicolson**
Fixes #540 

Review of colleague's PR #554

#### Changes proposed in this PR:

- implement bubble() function
- implement main function to call bubble with required input
- test main and bubble

### CODE REVIEW

File: 02_bubble/nicholsk/main.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	fmt.Fprintln(out, bubble([]int{3, 2, 1, 5}))
+}
+
+func bubble(s []int) []int {
+
+	// track whether the sort is complete
+	sorted := false
+
+	// make a slice copy to perform sort on
+	sCopy := s
```
Comment (mohitnag) [inline 02_bubble/nicholsk/main.go]: Not sure why you need a variable to make a copy over here.

File: 02_bubble/nicholsk/main.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+	for !sorted {
+
+		// track whether any swaps took place in current iteration
+		swaps := false
+
+		// iterate every index in the list, starting at index 1
+		for i := 1; i < len(s); i++ {
```
Comment (mohitnag) [inline 02_bubble/nicholsk/main.go:26]: why ```len(s)``` and why not ```len(sCopy)``` ?

**Comment by patrickmarabeas** [comment]
Don't forget to be descriptive with your commit messages and PR. Currently your PR does not inform the reviewer what has been done, nor the issue being resolved. Also remember to use the appropriate tags.

@mohitnag Don't forget to update the tags as well.

File: 02_bubble/nicholsk/main.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+)
+
+var out io.Writer = os.Stdout
+
+func main() {
+	fmt.Fprintln(out, bubble([]int{3, 2, 1, 5}))
+}
+
+func bubble(s []int) []int {
+
```
Comment (juliaogris) [inline 02_bubble/nicholsk/main.go:16]: Unnecessary blank line.
No need for blank lines at beginning or end of block.
Use blank lines between type, method and function definitions and within a block for logical separation.

File: 02_bubble/nicholsk/main.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	// track whether the sort is complete
+	sorted := false
```
Comment (juliaogris) [inline 02_bubble/nicholsk/main.go:18]: you don't need `sorted` and `swaps`; `swaps` alone should suffice.

File: 02_bubble/nicholsk/main.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+				// if less than previous then swap elements
+				s[i], s[i-1] = s[i-1], s[i]
+				swaps = true
+			}
+
+		}
+
+		// an iteration with no swaps indicates the array is sorted
+		if !swaps {
+			sorted = true
```
Comment (juliaogris) [inline 02_bubble/nicholsk/main.go:40]: return or break here

File: 02_bubble/nicholsk/main.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+		swaps := false
+
+		// iterate every index in the list, starting at index 1
+		for i := 1; i < len(s); i++ {
+
+			// compare element to previous
+			if s[i-1] > s[i] {
+
+				// if less than previous then swap elements
+				s[i], s[i-1] = s[i-1], s[i]
```
Comment (juliaogris) [inline 02_bubble/nicholsk/main.go:32]: you haven't created a copy of the input slice and you are there for modifying the input.
create a test case that fails when checking that input is left unmodified and fix with `copy()`

File: 02_bubble/nicholsk/main.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+func main() {
+	fmt.Fprintln(out, bubble([]int{3, 2, 1, 5}))
+}
+
+func bubble(s []int) []int {
+
+	// track whether the sort is complete
+	sorted := false
+
+	for !sorted {
```
Comment (juliaogris) [inline 02_bubble/nicholsk/main.go:20]: Typically bubble sort has to nested loops.
After the first iteration of the outer loop the largest element has "bubbled" up to the highest index position. 
Therefore no more comparisons with the highest index position need to be executed:

```go
	for i := len(s); i > 0; i-- {
		for j := 1; j < i; j++ {
			if s[j-1] > s[j] {
				s[j-1], s[j] = s[j], s[j-1]
			}
		}
	}
```
you can add in an early bail when nothing has been swapped ie the slice is already sorted:
```go
	for i := len(s); i > 0; i-- {
		swapped := false
		for j := 1; j < i; j++ {
			if s[j-1] > s[j] {
				s[j-1], s[j] = s[j], s[j-1]
				swapped = true
			}
		}
		if !swapped {
			// If no elements were swapped, the slice is sorted so we can stop here
			return
		}
	}
```

File: 02_bubble/nicholsk/main_test.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+	sorted := bubble([]int{3, 1, 2, 3, 2, 1})
+	expected := []int{1, 1, 2, 2, 3, 3}
+
+	a.Equal(expected, sorted)
+}
+
+func TestBubbleWithLargeSlice(t *testing.T) {
+	a := assert.New(t)
+
+	expected := make([]int, 100)
```
Comment (juliaogris) [inline 02_bubble/nicholsk/main_test.go:48]: Poor naming: this is the `input` not the `expectedOutput`.

File: 02_bubble/nicholsk/main_test.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+func TestBubbleWithSmallSliceIncludingNegativeInts(t *testing.T) {
+	a := assert.New(t)
+
+	sorted := bubble([]int{1, -2, -3})
+	expected := []int{-3, -2, 1}
+
+	a.Equal(expected, sorted)
+}
```
Comment (juliaogris) [inline 02_bubble/nicholsk/main_test.go:66]: Add test cases for empty slice and already sorted.

**Review by juliaogris — CHANGES_REQUESTED** [review]
Please [fix commit messages](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#commit-messages) with `git rebase -i HASH`.

Your bubble sort implementation is lacking a key parts to be called bubble sort IMO.

**Pull request closed without merging**